src/datanode/src/sql/insert.rs

```diff
@@ -123,6 +139,153 @@ impl SqlHandler {
+            }
+        );
+
+        let batch_schema = &batch.schema;
+        let batch_columns = batch_schema.column_schemas();
+        assert_eq!(batch_columns.len(), columns_num);
+        let mut columns_values = HashMap::with_capacity(columns_num);
+
+        for (i, column_name) in columns.into_iter().enumerate() {
+            let column_schema = schema
+                .column_schema_by_name(&column_name)
+                .with_context(|| ColumnNotFoundSnafu {
+                    table_name: &table_name,
+                    column_name: &column_name,
+                })?;
+            let expect_datatype = column_schema.data_type.as_arrow_type();
+            let batch_datatype = batch_columns[i].data_type.as_arrow_type();
+            let coerced_type = coerce_types(&expect_datatype, &Operator::Eq, &batch_datatype)
+                .map_err(|_| Error::ColumnTypeMismatch {
+                    column: column_name.clone(),
+                    expected: column_schema.data_type.clone(),
+                    actual: batch_columns[i].data_type.clone(),
+                })?;
+
+            ensure!(
+                expect_datatype == coerced_type,
+                ColumnTypeMismatchSnafu {
+                    column: column_name,
+                    expected: column_schema.data_type.clone(),
+                    actual: batch_columns[i].data_type.clone(),
+                }
+            );
+            let vector = batch
+                .column(i)
+                .cast(&column_schema.data_type)
+                .map_err(|_| Error::ColumnTypeMismatch {
+                    column: column_name.clone(),
+                    expected: column_schema.data_type.clone(),
+                    actual: batch_columns[i].data_type.clone(),
+                })?;
+
+            columns_values.insert(column_name, vector);
+        }
+
+        Ok(SqlRequest::Insert(InsertRequest {
+            catalog_name,
+            schema_name,
+            table_name,
+            columns_values,
+            region_number: 0,
+        }))
+    }
+
+    // FIXME(dennis): move it to frontend when refactor is done.
+    async fn build_stream_from_query(
+        &self,
+        table: TableRef,
+        stmt: Insert,
+        query_ctx: QueryContextRef,
+    ) -> Result<InsertRequstStream> {
+        let query = stmt
+            .query_body()
+            .context(ParseSqlValueSnafu)?
+            .context(MissingInsertBodySnafu)?;
+
+        let logical_plan = self
+            .query_engine
+            .statement_to_plan(
+                QueryStatement::Sql(Statement::Query(Box::new(query))),
+                query_ctx.clone(),
+            )
+            .context(ExecuteSqlSnafu)?;
+
+        let output = self
+            .query_engine
+            .execute(&logical_plan)
+            .await
+            .context(ExecuteSqlSnafu)?;
+
+        let batches = match output {
+            Output::Stream(s) => RecordBatches::try_collect(s)
```

Comment:
It seems you are collecting the recordbatch stream only to break it into another stream below in line 257. Can we just "map" the stream?

Reply:
Good catch!